### PR TITLE
Wait for ALB slow start during deploys

### PR DESCRIPTION
**Pull request by dustMason**
This change causes odin to pause long enough for any ALB slow start periods to elapse before tearing down outgoing ASGs.

Doing this avoids abrupt changes in load on some instances in the incoming ASG. For example, if 15 instances of a 30 instance ASG are still in their slow start period when the old ASG is torn down, the other 15 instances suddenly receive higher load, which drops down gradually as the slow start period elapses.

### CODE REVIEW

File: deployer/models/release.go [hunk 1]
```diff
@@ -29,7 +29,8 @@ type Release struct {
 	// Maintain a Log to look at what has happened
 	Healthy *bool `json:"healthy,omitempty"`
 
-	WaitForHealthy *int `json:"wait_for_healthy,omitempty"`
+	WaitForHealthy    *int `json:"wait_for_healthy,omitempty"`
+	PauseForSlowStart *int `json:"pause_for_slow_start,omitempty"`
```
Comment (grahamjenson) [inline deployer/models/release.go]: I think we should call this `WaitForDetach` in line with this other variable.

File: deployer/machine.go [hunk 2]
```diff
@@ -132,6 +132,11 @@ func StateMachine() (*machine.StateMachine, error) {
         ],
         "Default": "DetachForFailure"
       },
+      "WaitForDetach": {
+        "Type": "Wait",
+        "SecondsPath" : "$.pause_for_slow_start",
```
Comment (grahamjenson) [inline deployer/machine.go]: fix this?

File: deployer/models/release_resources.go [hunk 3]
```diff
@@ -122,13 +122,17 @@ func (release *Release) UpdateWithResources(resources *ReleaseResources) {
...
-func (release *Release) CreateResources(asgc aws.ASGAPI, cwc aws.CWAPI) error {
+func (release *Release) CreateResources(asgc aws.ASGAPI, cwc aws.CWAPI, albc aws.ALBAPI) error {
 	for _, service := range release.Services {
 		err := service.CreateResources(asgc, cwc)
 		if err != nil {
 			return err
 		}
 	}
+
+	d := release.SlowStartDuration(albc)
```
Comment (grahamjenson) [inline deployer/models/release_resources.go]: This needs to be in SetDefaults

File: deployer/models/service.go [hunk 4]
```diff
@@ -596,6 +601,22 @@ func (service *Service) UpdateHealthy(asgc aws.ASGAPI, elbc aws.ELBAPI, albc aws
 	return nil
 }
 
+func (service *Service) SlowStartDuration(albc aws.ALBAPI) int {
+	longest := 0
+	for _, arn := range service.Resources.TargetGroups {
+		output, _ := albc.DescribeTargetGroupAttributes(&elbv2.DescribeTargetGroupAttributesInput{TargetGroupArn: arn})
```
Comment (grahamjenson) [inline deployer/models/service.go]: check error.

File: deployer/models/service.go [hunk 5]
```diff
@@ -596,6 +601,22 @@ func (service *Service) UpdateHealthy(asgc aws.ASGAPI, elbc aws.ELBAPI, albc aws
...
+		for _, attribute := range output.Attributes {
+			if *attribute.Key == "slow_start.duration_seconds" {
```
Comment (grahamjenson) [inline deployer/models/service.go]: ensure attribute.Key and Value are not nil

Comment (dustMason) [reply deployer/models/service.go]: I dont think they can be nil, they are string pointers?

**Pull request merged**